File: server/src/data/ActiveGames.ts
import { setupAreaAssignments } from "../gameLogic/Controllers/AreaAssigner";
import { GameGenerator } from "../gameLogic/Controllers/GameGenerator";
import { Game } from "../gameLogic/Models/Game";
import { Player } from "../gameLogic/Models/Player";

class ActiveGames {
    private gamesInProgress: {[gameID: string]: Game};
    constructor() {
        this.gamesInProgress = {};
    }

    public createGame(userID: string, numberOfPlayers: number): Game {
        const areas = setupAreaAssignments(numberOfPlayers);
        const game = GameGenerator.generateGame(areas, numberOfPlayers, userID);
        this.gamesInProgress[game.getUUID()] = game;
        return game;
    }

    public getGameByID(id: string): Game {
        return this.gamesInProgress[id];
    }

    public getPlayerByUserID(gameID: string, userID: string): Player | null {
        const game = this.getGameByID(gameID);
        const players = game.getPlayers();

        for (const player of players) {
            if (player.getUserID() === userID) {
                return player;
            }
        }

        return null;
    }

    public deleteGame(id: string): void {
        delete this.gamesInProgress[id];
    }

    public hasGameStarted(gameID: string): boolean {
        const game = this.getGameByID(gameID);
        const players = game.getPlayers();

        for (let i = 0; i < players.length; i++) {
            if (players[i].getUserID() === '') {
                return false;
            }
        }

        return true;
    }

    public removeUserIDFromGame(userID: string, gameID: string): void {
        const game = this.getGameByID(gameID);
        const players = game.getPlayers();
    
        for (const player of players) {
            if (player.getUserID() === userID) {
                player.setUserID('');
            }
        }
    }
}

export const activeGames = new ActiveGames();

File: server/game_logic/ActiveGames.py
from game_logic.models.Game import Game
from game_logic.models.Player import Player
from controllers.game_setup import generate_game

class ActiveGames:
    def __init__(self):
        self.games_in_progress = {}

    def create_game(self, num_players: int, user_id: str) -> Game:
        game = generate_game(num_players, user_id)
        self.games_in_progress[game.uuid] = game
        return game
    
    def get_game_by_id(self, id: str) -> Game:
        return self.games_in_progress[id]
    
    def get_player_by_user_id(self, game_id: str, user_id: str) -> Player:
        game = self.games_in_progress[game_id]
        
        for player in game.players:
            if player.user_id == user_id:
                return player
            
    def delete_game(self, game_id: str):
        self.games_in_progress.pop(game_id)
        
    def has_game_started(self, game_id: str) -> bool:
        game = self.get_game_by_id(game_id)
        
        if not game:
            return False
        
        for player in game.players:
            if not player.user_id:
                return False 
            
        return True

    def remove_user_id_from_game(self, game_id: str, user_id: str):
        game = self.get_game_by_id(game_id)
        for player in game.players:
            if player.user_id == user_id:
                player.user_id = ""